### PR TITLE
Add documentation on Triple .NET Team Guidelines

### PR DESCRIPTION
**Pull request by Nuss9**
Adds a few documents on .NET team standards.

### CODE REVIEW

File: CodingGuidelines.md [hunk 1]
```diff
@@ -7,6 +7,10 @@ In order to improve team wide readablity and lower onboarding time we agreed to
 The coding guidelines can be found [here](https://csharpcodingguidelines.com/).
 Feel free to download the .PDF for later references.
 
+## IDE
+
+At Triple we are not tied to a single operating system, and so it happens that also among the .NET team there is cross-platform development. In addition to the free choice of IDE, there are quite a few flavors possible. However, in order to reduce startup and onboarding time, we should strive to facilitate cross-platform development as best as possible. This includes the generation of a README file, with instructions for environment specific setups. At the least, it should contain the instructions for the cross-platform editor  [Visual Studio Code](https://code.visualstudio.com/).
+
```
Comment (Nuss9) [inline CodingGuidelines.md:13]: Question: Should we make an exception for the `/.vscode/tasks.json` and `/.vscode/build.json` files to improve startup time?

Comment (ThomasBleijendaal) [reply CodingGuidelines.md:13]: Is it startup time or onboard time? Those files are generated right?

File: SourceControl.md [hunk 2]
```diff
@@ -0,0 +1,54 @@
+# General
+
+In the Triple .NET team we use Git for source control management. How you we [Git](https://git-scm.com) - whether it is via a GUI or the CLI - is up to us. Our commits however must stick to the guidelines:
```
Comment (Bas-kolthof-triple) [inline SourceControl.md:3]: "How you we Git", sentence is a bit broken here.

Comment (Bas-kolthof-triple) [inline SourceControl.md:3]: I'd rephrase this to 'is up to personal preference'

File: SourceControl.md [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+
+1. We must commit one change per commit
+2. We must adhere to the commit title and -description requirements
+3. We should not commit generated files
+4. We should not commit personal user settings
+5. We must include the .editorconfig in the root of your solution
+
+## Commit title and description
+
+We should strive to make our commits concise and descriptive, allowing to understand what happens within by only reading its title. By starting our commits with the following keywords we communicate our intent of the change in a uniform and succinct manner. In the optional description, you can find space to freely describe your commit in more detail.
```
Comment (Bas-kolthof-triple) [inline SourceControl.md:13]: Are these keywords from a standard somewhere? If so, reference it, if not, why these keywords specifically?

File: CodingGuidelines.md [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+## Testing
+
+We thorougly appreciate it when business logic is well covered by a test suite. This helps with extending functionality, refactoring existing code, and general readability.
+In order to promote productivity, we should use the [NUnit](https://nunit.org/) testing framework, in combination with the mocking framework [Moq](https://github.com/Moq) by default.
+
+### Further reading resources
+
+* [Clean Code series](https://www.pearson.com/us/higher-education/series/Robert-C-Martin-Series/348084.html) by Robert C. Martin
```
Comment (Bas-kolthof-triple) [inline CodingGuidelines.md:21]: The clean series is dated, and wasn't that great when it was recent.
I recommend not recommending it.

File: CodingGuidelines.md [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+## C# General
+
+In order to improve team wide readablity and lower onboarding time we agreed to stick to defined coding guidelines. These coding guidelines help us writing better and more coherent code, both on an individual level, as well as across individual developers since the style becomes more aligned.
+
+The coding guidelines can be found [here](https://csharpcodingguidelines.com/).
+Feel free to download the .PDF for later references.
+
+## IDE
+
+At Triple we are not tied to a single operating system, and so it happens that also among the .NET team there is cross-platform development. In addition to the free choice of IDE, there are quite a few flavors possible. However, in order to reduce startup and onboarding time, we should strive to facilitate cross-platform development as best as possible. This includes the generation of a README file, with instructions for environment specific setups. At the least, it should contain the instructions for the cross-platform editor  [Visual Studio Code](https://code.visualstudio.com/).
```
Comment (Bas-kolthof-triple) [inline CodingGuidelines.md:12]: Well, we try not to be tied to an OS, but in reality we are stuck with IIS for a lot (most? all?) of our projects, and that just doesn't run on linux as far as I know.  
I think we should consider recommending using windows as your development OS, but certainly not mandate it.

File: SourceControl.md [hunk 6]
```diff
@@ -0,0 +1,54 @@
+# General
+
+In the Triple .NET team we use Git for source control management. How you we [Git](https://git-scm.com) - whether it is via a GUI or the CLI - is up to us. Our commits however must stick to the guidelines:
+
+1. We must commit one change per commit
+2. We must adhere to the commit title and -description requirements
+3. We should not commit generated files
+4. We should not commit personal user settings
+5. We must include the .editorconfig in the root of your solution
```
Comment (Bas-kolthof-triple) [inline SourceControl.md:9]: Which .editorconfig? also, this isn't something related to commits, this is more of a project setup issue.

File: SourceControl.md [hunk 7]
```diff
@@ -0,0 +1,54 @@
+# General
+
+In the Triple .NET team we use Git for source control management. How you we [Git](https://git-scm.com) - whether it is via a GUI or the CLI - is up to us. Our commits however must stick to the guidelines:
+
+1. We must commit one change per commit
+2. We must adhere to the commit title and -description requirements
+3. We should not commit generated files
+4. We should not commit personal user settings
```
Comment (Bas-kolthof-triple) [inline SourceControl.md:8]: 'should'? are there valid reasons to sometimes commit user settings?

File: SourceControl.md [hunk 8]
```diff
@@ -0,0 +1,54 @@
+# General
+
+In the Triple .NET team we use Git for source control management. How you we [Git](https://git-scm.com) - whether it is via a GUI or the CLI - is up to us. Our commits however must stick to the guidelines:
+
+1. We must commit one change per commit
+2. We must adhere to the commit title and -description requirements
+3. We should not commit generated files
```
Comment (ThomasBleijendaal) [inline SourceControl.md:7]: In Umbraco generated files from database models should be commited as they are required. Generated migrations in Entity Framework should also be commited. Only like binaries or generated javascript files should not be commited, so this line should be changed to "Be careful when committing generated files" or something.

File: CodingGuidelines.md (hunk 1 above)
Comment (Nuss9) [reply CodingGuidelines.md:13]: I think we should make exceptions here when the (build) tasks deviate from the generated ones, to speed up the process of getting projects up and running (so from the online repo to actually running on one's machine).
In my current project I have an example, concerning setting up a host for an Azure function, and attaching the debugger to it.

File: SourceControl.md (hunk 3 above)
Comment (Nuss9) [reply SourceControl.md:13]: These keywords I've picked from online articles about best practices for committing. There's more available, and I've seen other keywords being used in the team, so let's open up the discussion!

File: SourceControl.md (hunk 7 above)
Comment (Nuss9) [reply SourceControl.md:8]: Yes, I'm referring here to CodingGuidelines.md > IDE. There's comments here that go into more detail about this issue - of which I'm not entirely sure yet.

File: SourceControl.md (hunk 8 above)
Comment (Nuss9) [reply SourceControl.md:7]: That's exactly why i used **should**. On the other side of this coin, there's also possibilities for code generation on the build server which would then be recommended over checking-in generated code into source control.

**Comment by ThomasBleijendaal** [comment]
@Nuss9 Can you either resolve or close the comments in this PR so I can merge it?

File: CodingGuidelines.md (hunk 5 above)
Comment (Nuss9) [reply CodingGuidelines.md:12]: So far I have worked on 3 different projects, all cross-platform ones. Seen the amount of new project we initiate, and the direction .NET is going into, we would do well anticipating this. 
True, older projects are still bound to Windows, and it would not be valuable to make these cross-platform, but for the newer ones we should keep this in mind - especially when onboarding external devs.

Comment (Nuss9) [reply CodingGuidelines.md:12]: However, I will implement the suggestion made in the meeting - if you work on a different OS and want to run it too, add to the readme how to initiate the project on your OS/IDE.

File: SourceControl.md (hunk 6 above)
Comment (Nuss9) [reply SourceControl.md]: Agreed, relocated this statement.

File: CodingGuidelines.md (hunk 4 above)
Comment (Nuss9) [reply CodingGuidelines.md:21]: I've gone through the [article](https://qntm.org/clean) you suggested why not to recommend _Clean Clode_. It has been 2 years now since I read the series and looking back at the few examples and arguments the author provides I have to admit that some of them are fuzzy indeed and the author has a point. 

Concerning the 'bad parts' in the book, I quote:
_"Is the whole book like this? Pretty much, yeah. Clean Code mixes together a disarming combination of strong, timeless advice and advice which is highly questionable or dated or both."_

I suggest focusing on the _'disarming combination of strong timeless advice'_. Let's not forget the author has had decades of hands-on experience in our field, which we should not casually dismiss. Let's instead read carefully and discuss to come to common ground, and use this to follow-up.

In the end, the author of the article found a resource worth recommending: [A Philosophy of Software Design (2018)](https://www.goodreads.com/en/book/show/39996759-a-philosophy-of-software-design). Have you read this? Accordingly, _"it focuses more on the practice of software design at a higher level than it does on the writing or critiquing of actual individual lines of code"_, so it would also be good to find a good resource on that. 
I will look into the book in order to make a good comparison.
In the comments we can find more examples of books recommended by readers. And seen the amount of them, there is no concensus on what to read, and what not. 

Therefore I think providing a reading list of resources we've enjoyed to suggest further reading is fine, as it opens discussion and we might learn a thing or two. Would you add your recommendations?

**Comment by Nuss9** [comment]
Resolved comments, ready to merge this PR. 
Further discussions or changes we can discuss in meetings, and submit via PR.
